### ISSUE TITLE
Aren't Class labels used during backpropagation ? 

### ISSUE THREAD

**Issue by eunjoa** [vlfeat/matconvnet#292]
Here is my problem, i am fine tuning the CIFAR model for my new dataset of only two classes, the labels are just 1 or 2. i want the output of my network to be probabilities for the two classes, i used softmax as a last layer so my output is of the from [1 1 2],   Now at backpropagation 

E = exp(bsxfun(@minus, X, max(X,[],3))) ;
L = sum(E,3) ;
Y = bsxfun(@rdivide, E, L) ;

if nargin <= 1, return ; end
Y = Y .* bsxfun(@minus, dzdY,  sum(dzdY .* Y, 3)) ;
where : dzdY = single(1) 

are the class labels not used in the backprop at the last layer ?? 
This is confusing me, i would appreciate any clarification.

**Comment by nightrome** [comment]
If you look at vl_simplenn, you will see that the softmaxloss is called as:
(forward pass)
res(i+1).x = vl_nnsoftmaxloss(res(i).x, l.class)

(backward pass)
res(i).dzdx = vl_nnsoftmaxloss(res(i).x, l.class, res(i+1).dzdx) ;

So the classes are definitely being used..

**Comment by eunjoa** [comment]
then if i use this code to test on a new image :+1:

net = load('net.mat') ;

im = imread('test.jpg') ;
im_ = single(im);
%im_ = im_ - net.normalization.averageImage; 

% run the CNN
res = vl_simplenn(net, im_) ;

% show the classification result
scores = squeeze(gather(res(end).x)) ;
[bestScore, best] = max(scores) ;
sprintf('%s (%d), score %.3f',net.classes.description{best}, best, bestScore) ;

i get an error at vl_simplenn when softmaxloss is called as there is no l.class for a new image.
what am i doing wrong?

**Comment by nightrome** [comment]
Of course. You didn't input any class labels!
Try to learn from the example in: https://github.com/vlfeat/matconvnet/blob/eab824b11ce55b0a349a632b198c44a83cec5f8e/examples/cnn_train.m

During training the layers are set as: net.layers{end}.class = labels
Also you need to make sure the last layer of the network is "softmaxloss", not "softmax"

At test time you don't need labels and the last layer needs to be softmax.

Note that I am not aware of the changes in the latest version. I.e. vl_nnsoftmaxloss is deprecated and now included in vl_nnloss..

**Comment by eunjoa** [comment]
thank you @nightrome i've been following the example in cnn_train,
at test time i am actually using net from epoch-30, that's what i am giving as an input to vl_simplenn(net,new_image);   net in this case has already softmaxloss as last layer. 
the way i am doing the test on a new image is wrong ?

**Comment by eunjoa** [comment]
![training](https://cloud.githubusercontent.com/assets/14211756/10663593/7e28cdde-78f7-11e5-9f47-f8bb2bdb1492.JPG)
 this is the result of training i am getting.

**Comment by nightrome** [comment]
As I said, at test time you want the last layer to be "softmax", not "softmaxloss". Just replace it with net.layers{end}.type = 'softmax'. Then run your test.
But your graphs also show that something went wrong with the error at training time, so you might want to check on that...

**Comment by eunjoa** [comment]
okay thank you i will try to fix that.
